File: frontend/src/types/portfolio.ts
export interface Portfolio {
  code: string;
  name: string;
  description?: string;
  status: string;
  started_at?: string;
  closed_at?: string;
  created_at?: string;
  updated_at?: string;
  /** 持仓明细二级分组维度覆盖（issue #144）：{"ASSET_STOCK": "style"}；
   * null/缺省 = 前端内置默认；仅存显式覆盖项 */
  display_config?: Record<string, string> | null;
  // 以下字段由后端在列表接口中动态计算返回
  total_value?: number;
  cumulative_return?: number;
  investor_count?: number;
  // 详情接口派生字段（issue #99）：最新快照总资产 / 累计收益（总资产 − 净投入）
  // 无快照（draft 等）时为 null
  total_profit?: number | null;
}

export interface PortfolioCreate {
  code: string;
  name: string;
  description?: string;
  display_config?: Record<string, string> | null;
}

export interface PortfolioUpdate {
  name?: string;
  description?: string;
  /** 显式传 null = 清空配置恢复默认；不传 = 不修改 */
  display_config?: Record<string, string> | null;
}

export interface PortfolioValueSnapshot {
  id: number;
  portfolio_code: string;
  total_value: number;
  total_shares: number;
  unit_price: number;
  snapshot_date: string;
  created_at?: string;
}

// GET /portfolios/{code}/nav-history 返回项
export interface NavHistoryRecord {
  snapshot_date: string;
  unit_price: number | null;
  total_value: number | null;
  total_shares: number | null;
}

// GET /portfolios/{code}/returns 返回（轻量口径：累计 + 年化）
export interface PortfolioReturns {
  portfolio_code: string;
  cumulative_return: number | null;
  annualized_return: number | null;
  initial_nav: number | null;
  current_nav: number | null;
  holding_days: number | null;
}

/**
 * GET /portfolios/{code}/performance 返回（全量绩效与风险指标）
 *
 * twr：时间加权收益率，消除资金进出影响（本系统为净值化记账，等于净值增长率）
 * mwr：资金加权收益率（XIRR），反映实际投入资金的年化回报
 * twr > mwr 通常说明大部分资金买在高位（加仓时点不佳）
 */
export interface PortfolioPerformance {
  portfolio_code: string;
  twr: number | null;
  twr_chained: number | null;
  annualized_twr: number | null;
  mwr: number | null;
  initial_nav: number | null;
  current_nav: number | null;
  holding_days: number | null;
  return_1m: number | null;
  return_3m: number | null;
  return_6m: number | null;
  return_ytd: number | null;
  return_1y: number | null;
  return_3y: number | null;
  max_drawdown: number | null;
  max_drawdown_peak_date: string | null;
  max_drawdown_trough_date: string | null;
  annualized_volatility: number | null;
  cash_flow_count: number;
  /** 两种 TWR 算法一致性自检：false 说明净值序列异常 */
  nav_series_consistent: boolean | null;
  /** 持有期 < 90 天时为 false：年化（尤其 MWR）属大幅外推，仅供参考 */
  annualization_reliable: boolean;
}

File: backend/app/schemas/portfolio.py
from pydantic import BaseModel
from typing import Dict, List, Optional
from datetime import date, datetime


class PortfolioBase(BaseModel):
    code: str
    name: str
    description: Optional[str] = None


class PortfolioCreate(PortfolioBase):
    # 持仓明细二级分组维度覆盖（issue #144）：{"ASSET_STOCK": "style", ...}，
    # 仅存显式覆盖项；校验以 asset_class_dimension_rule 规则表为准
    display_config: Optional[Dict[str, str]] = None


class PortfolioUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    # 显式传 null = 清空配置恢复默认；不传 = 不修改（exclude_unset 语义，
    # router 以 "display_config" in updates 区分后经哨兵传入 service）
    display_config: Optional[Dict[str, str]] = None


class PortfolioResponse(PortfolioBase):
    status: str = "draft"
    started_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # 组合级展示配置（issue #144）；ListItem 继承自动带出（列表页不消费）
    display_config: Optional[Dict[str, str]] = None
    # 读侧派生（详情接口并入，issue #99）：最新快照总资产 / 累计收益（总资产 − 净投入）
    # 无快照（draft 等）时为 None
    total_value: Optional[float] = None
    total_profit: Optional[float] = None

    class Config:
        from_attributes = True


class PortfolioListItem(PortfolioResponse):
    """列表项：附聚合字段（issue #69），无快照时为 None/0。"""
    total_value: Optional[float] = None
    cumulative_return: Optional[float] = None
    investor_count: int = 0


class PaginatedPortfolioResponse(BaseModel):
    items: List[PortfolioListItem]
    total: int
    page: int
    page_size: int


class PortfolioValueSnapshotResponse(BaseModel):
    id: int
    portfolio_code: str
    snapshot_date: date
    total_value: float
    total_shares: float
    unit_price: float
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PortfolioInvestorItem(BaseModel):
    investor_code: str
    name: str
    shares: float


class NavHistoryRecord(BaseModel):
    snapshot_date: date
    unit_price: Optional[float] = None
    total_value: Optional[float] = None
    total_shares: Optional[float] = None


class PortfolioPerformance(BaseModel):
    """组合绩效指标（快照不足时相应字段为 None）。

    twr：时间加权收益率，消除资金进出影响（净值化系统下等于净值增长率）
    mwr：资金加权收益率（XIRR），反映实际投入资金的年化回报
    """
    portfolio_code: str
    # 收益率
    twr: Optional[float] = None
    twr_chained: Optional[float] = None
    annualized_twr: Optional[float] = None
    mwr: Optional[float] = None
    # 净值与持有期
    initial_nav: Optional[float] = None
    current_nav: Optional[float] = None
    holding_days: Optional[int] = None
    # 区间收益（组合成立不足窗口期时为 None）
    return_1m: Optional[float] = None
    return_3m: Optional[float] = None
    return_6m: Optional[float] = None
    return_ytd: Optional[float] = None
    return_1y: Optional[float] = None
    return_3y: Optional[float] = None
    # 风险指标
    max_drawdown: Optional[float] = None
    max_drawdown_peak_date: Optional[str] = None
    max_drawdown_trough_date: Optional[str] = None
    annualized_volatility: Optional[float] = None
    # 元信息
    cash_flow_count: int = 0
    nav_series_consistent: Optional[bool] = None
    # 持有期 < 90 天时为 False：年化属大幅外推，前端应标注仅供参考
    annualization_reliable: bool = False
